Don't crash in the presence of non-UTF-8 env vars.

benjyw opened the pull request:
Closes https://github.com/pantsbuild/pants/issues/19199

benjyw commented on src/rust/engine/src/externs/nailgun.rs:
```diff
@@ -44,6 +44,8 @@ impl PyNailgunClient {
   fn execute(&self, command: String, args: Vec<String>, env: &PyDict, py: Python) -> PyResult<i32> {
     use nailgun::NailgunClientError;
 
+    // NB: We assume that env var names and values are strict UTF-8 (with no lone surrogates
```
It was substantially easier to push this responsibility to the Python side rather than verify this on the Rust side, which AFAICT would require a lot of mucking about with the implementation of `FromPyObject::extract`, and adding overhead even in the normal case where we are justified in expecting valid strings.  Instead we say "anything you expect to convert to a Rust String must be a Python str that is strictly encodable to UTF-8." LMK if you think we should do the hoop-jumping in Rust.

benjyw commented on src/rust/engine/src/externs/nailgun.rs:
```diff
@@ -44,6 +44,9 @@ impl PyNailgunClient {
   fn execute(&self, command: String, args: Vec<String>, env: &PyDict, py: Python) -> PyResult<i32> {
     use nailgun::NailgunClientError;
 
+    // NB: We assume that env var names and values are Python strs strictly convertible to UTF-8
```
It was substantially easier to push this responsibility to the Python side rather than verify this on the Rust side, which AFAICT would require a lot of mucking about with the implementation of `FromPyObject::extract`, and adding overhead even in the normal case where we are justified in expecting valid strings.  Instead we say "anything you expect to convert to a Rust String must be a Python str that is strictly encodable to UTF-8." LMK if you think we should do the hoop-jumping in Rust.

benjyw commented on src/rust/engine/client/src/main.rs:
```diff
@@ -53,7 +53,10 @@ enum PythonLogLevel {
 async fn execute(start: SystemTime) -> Result<i32, String> {
   let build_root = BuildRoot::find()?;
-  let options_parser = OptionParser::new(Env::capture(), Args::argv())?;
+  let env = Env::capture_lossy();
```
Rather than capture the Env twice, as before, we re-use this Env for execute_command. So we must capture the items and argv before moving it into options_parser.

benjyw replied:
The API here is a bit idiosyncratic - OptionParser wants a HashMap, execute_command wants a Vec of key-value pairs, but that was more than I wanted or needed to unravel, hence the `From` implementation to get that Vec.

jsirois commented on src/python/pants/bin/remote_pants_runner.py:
```diff
@@ -132,14 +132,21 @@ def _connect_and_execute(
+        #  as lone surrogates (see https://peps.python.org/pep-0383/).
+        #  However when we pass these to Rust, we will fail to decode as strict UTF-8.
+        #  So we perform a lossy re-encoding to prevent this.
+        def strict_utf8(s: str) -> str:
```
Stepping away back, the ideal is to be lazy. This sort of conversion is not correct. You can imagine, if the env var pointed to a path for example, it now no longer does. For the scie side of these fixes, that's what I did. Only force the env vars you actually need (read and act on) to be utf-8. For those you don't care about, definitely don't muck since you know not what you may have broken.

jsirois replied:
My memory was incorrect on how I fixed. I am lazy, but I warm, which doesn't help someone that really needs to deal with non-utf-8: https://github.com/a-scie/jump/pull/106#discussion_r1335039559

This stuff is hard to do right but at least the Rust OsString/OsStr apis keep you honest.

benjyw replied:
Yeah, good point. My purpose here was not to support non-UTF-8 in env vars we consume, but to prevent unrelated env vars that we don't even consume from crashing us.  Possibly this should WARN though!

Actually supporting non-UTF-8 would be more work, as a lot of types will have to change in various places, so I'm unsure if it's worth it. I can take a stab at it on top of this though.

jsirois replied:
Yeah, you saved a Pants crash but caused a tool to fail silentlyish by corrupting the env vars it uses. I agree that the window of env vars you'd corrupt that a tool actually uses may be small, but warning may be enough to allow the person in that narrow corner to see the issue and call home and complain that, for them, this is a bug that needs to be fixed.

benjyw replied:
OK, fixed this up to WARN on such env vars, and ignore them. I think that's a better place to be than our current inability to run Pants at all even in the (likely?) case that the non-utf8 env var is not even consumed by Pants or any of its tools.

jsirois replied:
Yeah, that's better. I don't agree with Huon that the user who hits the issue files a feature request - this is a straight up bug from day one, but this is better and feature / bug is a pedantic distinction at that point anyhow - it just places the pressure in the right spot I think.

huonw commented on src/rust/engine/options/src/env.rs:
```diff
@@ -19,8 +19,29 @@ impl Env {
+    let mut env: HashMap<String, String> = HashMap::with_capacity(env_os.size_hint().0);
+    let mut dropped: Vec<String> = Vec::new();
+    for (os_key, os_val) in env_os {
+      match os_key.into_string() {
+        Ok(key) => match os_val.into_string() {
+          Ok(val) => {
+            env.insert(key, val);
+          }
+          Err(_) => {
+            dropped.push(key);
+          }
+        },
+        Err(os_key) => {
+          // We'll only be able to log the lossy name of any non-UTF-8 keys, but
+          // the user will know which one we mean.
+          dropped.push(os_key.to_string_lossy().to_string());
+        }
+      }
```
FYI, just stylistically for Rust, this could expressed as a match on a tuple. This makes the `os_val.into_string()` happen more often, but the "hot"/common path is running both `into_string`s anyway, so I think that's fine... It also makes it further from the Python code 🤷‍♂️ 

```suggestion
      match (os_key.into_string(), os_val.into_string()) {
        (Ok(key), Ok(val)) => { env.insert(key, val); },
        (Ok(key), Err(_)) => dropped.push(key),
        // We'll only be able to log the lossy name of any non-UTF-8 keys, but
        // the user will know which one we mean.
        (Err(os_key), _) => dropped.push(os_key.to_string_lossy().to_string()),
      }
```

(I suspect that phrasing doesn't pass `cargo fmt`)

huonw reviewed (approved):
Dropping env vars (noisily) makes sense as a major improvement here, and people can file feature requests for supporting non-UTF-8 paths in options (or whatever's required), if people encounter them.

Nice one!

jsirois commented on src/rust/engine/client/src/main.rs:
```diff
@@ -82,10 +85,12 @@ async fn execute(start: SystemTime) -> Result<i32, String> {
 
+  // Now that the logger has been set up, we can retroactively log any dropped env vars.
+  for name in dropped {
+    log::warn!("Environment variable with non-UTF-8 name or value ignored: {name}");
```
It seems like it would have been easy to keep name and value (as OsString) and just log both here. It seems slightly mean to not log the value in the case where the value was the issue - which, IIUC, is the only actual known issue case so far.

benjyw replied:
My concern with that was that the value may be unprintable in practice (e.g., it could be several K of binary data). We already know we're on thin ice trying to print it here. The name is far more likely to be short and (close to) printable. I suppose we could truncate it. But either way I think the name is what matters in practice for remediating the problem.

jsirois reviewed (approved):
It would be great to have a test for these fixes.

benjyw commented:
OK:

1. Switched to Huon's suggested pairwise match.
2. Added separate logging for bad keys vs keys with bad values.
3. A test (and slight refactoring of env.rs to enable to the test without mucking with the real environment).

PTAL

jsirois commented on src/rust/engine/options/src/env.rs:
```diff
@@ -14,13 +15,46 @@ pub struct Env {
+          env.insert(key, val);
+        }
+        (Ok(key), Err(_)) => dropped.keys_with_non_utf8_values.push(key),
+        // We'll only be able to log the lossy name of any non-UTF-8 keys, but
```
This is a non-local comment - no logging here. I think it'd be better to just store `OsString` in `DroppedEnvVars` and let the user decide what to do in the context of their use. They can convert to String and sort if they so desire or just print OsString debug content, or - whatever. They have everything they need.

jsirois commented on src/rust/engine/options/src/env_tests.rs:
```diff
@@ -15,6 +16,39 @@ fn env<I: IntoIterator<Item = (&'static str, &'static str)>>(vars: I) -> Env {
 
+#[test]
+#[cfg(not(target_os = "windows"))]
+fn test_capture_lossy() {
```
Ok. This sort of test is a bit tautological, I was thinking an IT to actually show you could run pants in an env with bad keys and values, but it doesn't hurt too much.

benjyw replied:
I don't think it's tautological - it tests that we correctly classify env vars, and don't crash. True though that it does assume that this is the production pathway for reading env vars.

Still, I also added an IT.

benjyw replied:
(and verified that it fails in the expected way on main)

huonw commented on src/python/pants/bin/loader_integration_test.py:
```diff
@@ -82,3 +82,16 @@ def run(limit: str) -> PantsResult:
+        },
+        use_pantsd=True,
+    )
+    print(res.stdout)
```
Is this `print` left over? And/or are we intending to validate that the output references the problematic variables?

benjyw replied:
Ooops, yes. Debug print.

huonw reviewed (approved):
Nice!

jsirois commented on src/rust/engine/options/src/env.rs:
```diff
@@ -14,13 +15,42 @@ pub struct Env {
+        (Err(os_key), _) => dropped.non_utf8_keys.push(os_key),
+      }
+    }
+    dropped.non_utf8_keys.sort();
```
I think any form of display considerations being mixed here is not a great idea. The OsString -> String conversion for display is now gone, which I think is good, but I'd take it 1 step further is you actually agree with the reasoning here. It either all applies or it doesn't it seems to me.

benjyw replied:
Do you mean to not sort here? I'd prefer to keep it. It's potentially useful to the caller (and ensures consistent output in tests, so it's not just a display consideration), and at worst does no harm. If the caller wants to sort they have to make a mut copy and sort it, which is a couple of extra lines, and more allocation, for each vec. Seems unnecessary.

jsirois replied:
Your last sentence is untrue in terms of copying. For example, this works and prints `a z`:
```rust
fn main() {
    pub struct Foo {
        pub bar: Vec<String>
    }
    let f = Foo { bar: vec!["z".into(), "a".into()] };
    let mut b = f.bar;
    b.sort();
    println!("{bar}", bar = b.join(" "));
}
```

I find it strange to do a 3rd party sort. The sort isn't built into the contract of the struct right now; so some are sorted, some not. If your claims about sorting being needed for test reasons, then that's obviously valid. It doesn't ring true to me, but your head is in the code.

benjyw replied:
So... a couple of extra lines, as I said... True, no extra allocation if you're ok moving the Vec, and making the struct subsequently unusable. Which I guess we are in our current usage of this struct.

Still, since I don't really feel strongly about it, I will remove the sorts here and push them to the caller.

jsirois replied:
Excellent. Of the 3, caller decides, struct guarantees, 3rdparty does it I think the way you had it was the worst possible.

jsirois replied:
> and making the struct subsequently unusable.

You take the example too literally. You can take the whole struct as mut in the reciever and mutate one of its field in-place.

jsirois replied:
A more on-point example that works:
```rust
fn main() {
    pub struct Foo {
        pub bar: Vec<String>
    }
    fn example() -> (u64, Foo) {
        (42, Foo { bar: vec!["z".into(), "a".into()] })
    }
    let (mol, mut f) = example();
    f.bar.sort();
    println!("{mol}: {bar}", bar = f.bar.join(" "));
}
```
